src/python/pants/engine/target.py

```diff
@@ -1269,65 +1314,175 @@ def filespec(self) -> Filespec:
             args=includes, exclude=[excludes], root=self.address.spec_path
         )
 
+    @final
+    @classmethod
+    def can_generate(cls, output_type: Type["Sources"], union_membership: UnionMembership) -> bool:
+        """Can this Sources field be used to generate the output_type?"""
```

Comment:
Adding this method is important. While we generally expect people to rely on `HydratedSourcesRequest.for_sources_types` to know if codegen is possible, there are some cases where we need to pre-filter without hydrating the field. For example, `run_setup_py.py` needs this.

This method allows us to fully support the pre-filtering idiom that we normally use with the Target API via `tgt.has_field()`.

Reply:
The docstring here could maybe indicate that most callers won't need to call this, and recommend `HydratedSourcesRequest`.